File: src/main/java/mnm/mods/tabbychat/util/Color.java
package mnm.mods.tabbychat.util;

import com.google.gson.annotations.SerializedName;

import java.util.Random;

/**
 * Represents a color and provides an easy way to convert to and from html color
 * codes. A few default colors have been provided.
 */
public class Color {

    public static final Color BLACK = of(0xff000000);
    public static final Color DARK_BLUE = of(0xff0000aa);
    public static final Color DARK_GREEN = of(0xff00aa00);
    public static final Color DARK_AQUA = of(0xff00aaaa);
    public static final Color DARK_RED = of(0xffaa0000);
    public static final Color DARK_PURPLE = of(0xffaa00aa);
    public static final Color GOLD = of(0xffffaa00);
    public static final Color GRAY = of(0xffaaaaaa);
    public static final Color DARK_GRAY = of(0xff555555);
    public static final Color BLUE = of(0xff5555ff);
    public static final Color GREEN = of(0xff55ff55);
    public static final Color AQUA = of(0xff55ffff);
    public static final Color RED = of(0xffff5555);
    public static final Color LIGHT_PURPLE = of(0xffff55ff);
    public static final Color YELLOW = of(0xffffff55);
    public static final Color WHITE = of(0xffffffff);

    private static Random random = new Random();

    @SerializedName("r")
    private final int red;
    @SerializedName("g")
    private final int green;
    @SerializedName("b")
    private final int blue;
    @SerializedName("a")
    private final int alpha;

    /**
     * Creates a color using an html color code.
     *
     * @param hexColor The html color code
     */
    private Color(int hexColor) {
        this.alpha = hexColor >> 24 & 255;
        this.red = hexColor >> 16 & 255;
        this.green = hexColor >> 8 & 255;
        this.blue = hexColor & 255;
    }

    /**
     * Creates a color using RGBA color format.
     *
     * @param red Red color
     * @param green Green color
     * @param blue Blue color
     * @param alpha Transparency
     */
    private Color(int red, int green, int blue, int alpha) {
        this.red = red % 256;
        this.green = green % 256;
        this.blue = blue % 256;
        this.alpha = alpha % 256;
    }

    /**
     * Gets the html hex color code of this color usable by Minecraft's
     * {@link net.minecraft.client.gui.Gui}.
     *
     * @return The html hex code.
     */
    public int getHex() {
        int alphaI = alpha << 24;
        int redI = red << 16;
        int greenI = green << 8;
        return alphaI + redI + greenI + blue;
    }

    /**
     * Gets the red value.
     *
     * @return The red value
     */
    public int getRed() {
        return red;
    }

    /**
     * Gets the green value.
     *
     * @return The green value
     */
    public int getGreen() {
        return green;
    }

    /**
     * Gets the blue value.
     *
     * @return The blue value
     */
    public int getBlue() {
        return blue;
    }

    /**
     * Gets the alpha value.
     *
     * @return The alpha value
     */
    public int getAlpha() {
        return alpha;
    }

    /**
     * Convenience method for getting the html code for a RGBA color. Creates a
     * new object and calls {@link Color#getHex()}.
     *
     * @param red The red value
     * @param green The green value
     * @param blue The blue value
     * @param alpha The alpha value
     * @return The hexadecimal representation of this color.
     */
    public static int getColor(int red, int green, int blue, int alpha) {
        return of(red, green, blue, alpha).getHex();
    }

    /**
     * Generates a random color.
     *
     * @return A random color
     */
    public static Color random() {
        return of(random.nextInt());
    }

    public static Color of(int hexColor) {
        return new Color(hexColor);
    }

    public static Color of(int red, int green, int blue, int alpha) {
        return new Color(red, green, blue, alpha);
    }

    @Override
    public int hashCode() {
        final int prime = 31;
        int result = 1;
        result = prime * result + alpha;
        result = prime * result + blue;
        result = prime * result + green;
        result = prime * result + red;
        return result;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj)
            return true;
        if (obj == null || !(obj instanceof Color))
            return false;
        Color other = (Color) obj;
        return alpha == other.alpha
                && blue == other.blue
                && green == other.green
                && red == other.red;
    }
}

File: src/main/kotlin/mnm/mods/tabbychat/util/Color.kt
package mnm.mods.tabbychat.util

import net.minecraft.util.text.TextFormatting

/**
 * Represents a color and provides an easy way to convert to and from html color
 * codes. A few default colors have been provided via the [Colors] object.
 */
typealias Color = Int

/**
 * The color's alpha value
 */
val Color.alpha get() = this shr 24 and 255

/**
 * The color's red value
 */
val Color.red get() = this shr 16 and 255

/**
 * The color's green value
 */
val Color.green get() = this shr 8 and 255

/**
 * The color's blue value
 */
val Color.blue get() = this and 255

/**
 * The color's string represenation.
 */
val Color.asString get() = "#" + this.asLong.toString(16).padStart(8, '0')

/**
 * The unsigned int value as a long. Will always be positive
 */
val Color.asLong get() = Integer.toUnsignedLong(this)

fun rgb(r: Int, g: Int, b: Int) = argb(255, r, g, b)

fun argb(a: Int, r: Int, g: Int, b: Int): Color {
    var c = 0
    c += ((a and 255) shl 24)
    c += ((r and 255) shl 16)
    c += ((g and 255) shl 8)
    c += ((b and 255))
    return c
}

fun color(c: Long): Color {
    return c.toInt()
}

fun color(s: String): Color? {
    // first, try to parse it as a hex.
    // if it's not a hex, it must be a name!
    return parseColorHex(s) ?: getVanillaColorByName(s)
}

private fun parseColorHex(s: String): Color? {
    val reg = Regex("""^#([a-f0-9]+)$""", RegexOption.IGNORE_CASE)
    val match = reg.find(s)
    if (match != null && s.length in intArrayOf(4, 7, 9)) {
        val hex = StringBuilder(match.groupValues[1])
        if (hex.length == 3) {
            hex.insert(2, hex[2])
            hex.insert(1, hex[1])
            hex.insert(0, hex[0])
        }
        if (hex.length == 6) {
            hex.insert(0, "ff")
        }
        val (alpha, red, green, blue) = hex.chunked(2) {
            it.toString().toInt(16)
        }
        return argb(alpha, red, green, blue)
    }
    return null
}

private fun getVanillaColorByName(s: String): Color? {
    return TextFormatting.getValueByName(s)
            // with full opacity
            ?.color?.with(a = 255)
}

/**
 * Color Function
 */
typealias CF = (Int) -> Int

fun Color.with(a: Int = alpha, r: Int = red, g: Int = green, b: Int = blue): Color {
    return with({ a }, { r }, { g }, { b })
}

fun Color.with(a: CF = { it }, r: CF = { it }, g: CF = { it }, b: CF = { it }): Color {
    return argb(a(alpha), r(red), g(green), b(blue))
}

object Colors {
    val BLACK = color("black")!!
    val DARK_BLUE = color("dark_blue")!!
    val DARK_GREEN = color("dark_green")!!
    val DARK_AQUA = color("dark_aqua")!!
    val DARK_RED = color("dark_red")!!
    val DARK_PURPLE = color("dark_purple")!!
    val GOLD = color("gold")!!
    val GRAY = color("gray")!!
    val DARK_GRAY = color("dark_gray")!!
    val BLUE = color("blue")!!
    val GREEN = color("green")!!
    val AQUA = color("aqua")!!
    val RED = color("red")!!
    val LIGHT_PURPLE = color("light_purple")!!
    val YELLOW = color("yellow")!!
    val WHITE = color("white")!!
}
